core/src/main/java/org/apache/carbondata/core/datamap/status/DiskBasedDataMapStatusProvider.java

```diff
@@ -0,0 +1,173 @@
+ * limitations under the License.
+ */
+
+package org.apache.carbondata.core.datamap.status;
+
+import java.io.*;
+import java.nio.charset.Charset;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import org.apache.carbondata.common.logging.LogService;
+import org.apache.carbondata.common.logging.LogServiceFactory;
+import org.apache.carbondata.core.constants.CarbonCommonConstants;
+import org.apache.carbondata.core.datastore.impl.FileFactory;
+import org.apache.carbondata.core.fileoperations.AtomicFileOperations;
+import org.apache.carbondata.core.fileoperations.AtomicFileOperationsImpl;
+import org.apache.carbondata.core.fileoperations.FileWriteOperation;
+import org.apache.carbondata.core.locks.CarbonLockFactory;
+import org.apache.carbondata.core.locks.CarbonLockUtil;
+import org.apache.carbondata.core.locks.ICarbonLock;
+import org.apache.carbondata.core.locks.LockUsage;
+import org.apache.carbondata.core.metadata.schema.table.DataMapSchema;
+import org.apache.carbondata.core.util.CarbonProperties;
+import org.apache.carbondata.core.util.CarbonUtil;
+
+import com.google.gson.Gson;
+
+public class DiskBasedDataMapStatusProvider implements DataMapStatusStorageProvider {
+
+  private static final LogService LOG =
+      LogServiceFactory.getLogService(DiskBasedDataMapStatusProvider.class.getName());
+
+  private static final String DATAMAP_STATUS_FILE = "datamapstatus";
+
+  @Override public DataMapStatusDetail[] getDataMapStatusDetails() throws IOException {
+    String statusPath = CarbonProperties.getInstance().getSystemFolderLocation()
+        + CarbonCommonConstants.FILE_SEPARATOR + DATAMAP_STATUS_FILE;
+    Gson gsonObjectToRead = new Gson();
+    DataInputStream dataInputStream = null;
+    BufferedReader buffReader = null;
+    InputStreamReader inStream = null;
+    DataMapStatusDetail[] dataMapStatusDetails;
+    try {
+      if (!FileFactory.isFileExist(statusPath)) {
+        return new DataMapStatusDetail[0];
+      }
+      dataInputStream =
+          FileFactory.getDataInputStream(statusPath, FileFactory.getFileType(statusPath));
+      inStream = new InputStreamReader(dataInputStream,
+          Charset.forName(CarbonCommonConstants.DEFAULT_CHARSET));
+      buffReader = new BufferedReader(inStream);
+      dataMapStatusDetails = gsonObjectToRead.fromJson(buffReader, DataMapStatusDetail[].class);
+    } catch (IOException e) {
+      LOG.error(e, "Failed to read datamap status");
+      throw e;
+    } finally {
+      CarbonUtil.closeStreams(buffReader, inStream, dataInputStream);
+    }
+
+    // if dataMapStatusDetails is null, return empty array
+    if (null == dataMapStatusDetails) {
+      return new DataMapStatusDetail[0];
+    }
+
+    return dataMapStatusDetails;
+  }
+
+  @Override
+  public void updateDataMapStatus(List<DataMapSchema> dataMapSchemas, DataMapStatus dataMapStatus)
+      throws IOException {
+    ICarbonLock carbonTableStatusLock = getDataMapStatusLock();
```

Comment:
add some comment to tell it is going to overwrite the status file

Reply:
ok